src/main/java/uk/gov/hmcts/ccd/data/casedetails/search/SortOrderQueryBuilder.java

```diff
@@ -12,23 +21,39 @@
     private static final String SPACE = " ";
     private static final String COMMA = ",";
 
+    private static final Set<String> VALID_COLUMNS_FOR_ORDER_BY
+        = Collections.unmodifiableSet((Set<? extends String>) Stream
+                                          .of(CaseDetailsEntity.CREATED_DATE_FIELD_COL,
+                                              CaseDetailsEntity.LAST_MODIFIED_FIELD_COL,
+                                              CaseDetailsEntity.JURISDICTION_FIELD_COL,
+                                              CaseDetailsEntity.CASE_TYPE_ID_FIELD_COL,
+                                              CaseDetailsEntity.STATE_FIELD_COL,
+                                              CaseDetailsEntity.REFERENCE_FIELD_COL,
+                                              CaseDetailsEntity.SECURITY_CLASSIFICATION_FIELD_COL)
+                                          .collect(Collectors.toCollection(HashSet::new)));
```

Comment:
Based on the latest discussions around logic duplication between def store and data store, I think we can indeed remove this allowed-list approach as well. We can stay within the purpose of just SQL-i prevention by just rejecting a few special chars in case field ids, and indeed, case field IDs can reasonable assumed to not contain such chars, no matter how their validity conventions change in the foreseeable future.

Reply:
The [OWASP guide](https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/SQL_Injection_Prevention_Cheat_Sheet.md) recommends blacklisting only as a last resort:

> This technique should only be used as a last resort, when none of the above are feasible. Input validation is probably a better choice as this methodology is frail compared to other defenses and we cannot guarantee it will prevent all SQL Injection in all situations.

If in this case we know the range of possible values then Andrew's original whitelisting would be my preference but either approach is an improvement.

Regarding the duplication between definition and data stores, I see this as an indicator that these services should be merged together, which would also simplifying testing (no wiremock or duplication of models, cannot commit breaking changes to ccd into def store).

Reply:
Yes I was following the OWASP guide and initially tried to do white listing solution. The problem with that is it creates dependency on whatever change is introduced to `case_data` table's columns. I am ok with either solution here but the the common denominator between @MSancaktutar @smathangi and myself was blacklisting...
@MSancaktutar let me know what you think?

Reply:
Sanitising input against a disallowed list is in this case more efficient and well-secure. It is an ORDER BY clause we're building and preventing those disallowed chars solves the same problem to the same effect with less coupling with other parts of CCD.

Reply:
The OWASP guide is explicit that blacklisting should be the last resort and only used for securing legacy code; there are issues such as Unicode smuggling where special characters can get past a blacklist and get converted to dangerous characters by the database.

In this case the range of valid input is known so whitelisting is the secure way to go; a switch statement that only outputs known valid constants.

Database queries are inherently coupled to the schema they operate on, which is why it's problematic to split schema out of a database into its own service as we have done. If we are going to persist with the split then duplication is inevitable.

Reply:
I think this whitelist check for metadata fields are already happening against a ENUM here   https://github.com/hmcts/ccd-data-store-api/blob/75fbac732d9e24f220afa20f4f56f487f5c5a03c/src/main/java/uk/gov/hmcts/ccd/data/casedetails/search/SortOrderQueryBuilder.java#L42

Reply:
Yes @smathangi this is the right approach, a couple more places where the same needs to be done.

Reply:
@banderous can you let me know which places do you mean?

Reply:
@andrewfolga individual comments are in the review I made requesting changes.